app/src/main/java/com/fstyle/structure_android/utils/Validator.java

```diff
@@ -0,0 +1,359 @@
+                Object real = field.get(object);
+                // detect when data has changed only if onlyValidateTheChange is true.
+                // Otherwise, always run validation on all of fields .
+                if (!onlyValidateChange || (cache == null && real != null) || (cache != null
+                        && cache.equals(real))) {
+                    boolean valid = validate(real, rules, isOptional);
+                    if (!valid) {
+                        isValid = false;
+                    }
+                }
+            } catch (IllegalAccessException e) {
+                Log.e(TAG, "validate: ", e);
+            }
+            try {
+                if (model != null) {
+                    mModelCache = model.clone();
+                }
+                mViewCache = view;
+            } catch (CloneNotSupportedException e) {
+                mModelCache = null;
+                mViewCache = null;
+                Log.e(TAG, "validate: ", e);
+            }
+        }
+        return isValid;
+    }
+
+    public <T extends BaseModel> boolean validateAll(T model, boolean onlyValidateChange) {
+        return validateAll(model, null, onlyValidateChange);
+    }
+
+    public <V extends BaseView> boolean validateAll(V view, boolean onlyValidateChange) {
+        return validateAll(null, view, onlyValidateChange);
+    }
+
+    public <T extends BaseModel> Validator prepare(T model) {
+        try {
+            mModelCache = model.clone();
+        } catch (CloneNotSupportedException e) {
+            Log.e(TAG, "prepare: ", e);
+        }
+        return this;
+    }
+
+    @StringRes
+    public int getErrorByTypeInt(@ValidType int type) {
+        int index = mAllErrorMessage.indexOfValue(type);
+        if (index == -1) {
+            return R.string.empty;
+        }
+        return mAllErrorMessage.keyAt(index);
+    }
+
+    public String getErrorByType(@ValidType int type) {
+        int index = mAllErrorMessage.indexOfValue(type);
+        if (index == -1 || mAllErrorMessage.keyAt(index) == -1) {
+            return null;
+        }
+        return mContext.getString(mAllErrorMessage.keyAt(index));
+    }
+
+    public void initNGWordPattern() {
+        Observable.create(new Observable.OnSubscribe<Pattern>() {
+            @Override
+            public void call(Subscriber<? super Pattern> subscriber) {
+                try {
+                    BufferedReader reader = new BufferedReader(
+                            new InputStreamReader(mContext.getAssets().open("ng-word")));
+                    StringBuffer buffer = new StringBuffer();
+                    String line;
+                    while ((line = reader.readLine()) != null) {
+                        if (buffer.length() == 0) {
+                            buffer.append("(").append(line.toLowerCase(Locale.ENGLISH));
+                        } else {
+                            buffer.append("|").append(line.toLowerCase(Locale.ENGLISH));
```

Comment:
WARNING: [AppendCharacterWithChar] Avoid appending characters as strings in StringBuffer.append.
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/strings.html#AppendCharacterWithChar